tripy/nvtripy/frontend/module/module.py

```diff
@@ -37,26 +37,30 @@ def _check_param_compatible(original_param, new_param, param_name):
             f"Expected a tensor for parameter: '{param_name}', but got: {type_str_from_arg(new_param)}",
         )
 
-    if not isinstance(original_param, Tensor) and not isinstance(original_param, DefaultParameter):
+    if not isinstance(original_param, (Tensor, DefaultParameter, OptionalParameter)):
         # Allow values to be initialized to non-tensor types and changed later.
         # Note that this is required for the constructor to work since `original_param` will not be set.
         return
 
     is_compatible = utils.result.Result.ok()
 
-    # We need to evaluate here anyway, so we map the entire shape to numbers upfront to save us from recomputing
-    # them again later.
-    original_shape = tuple(map(int, original_param.shape))
-    new_shape = tuple(map(int, new_param.shape))
-    if original_shape != new_shape:
-        is_compatible = utils.result.Result.err(
-            ["New parameter shape: ", new_shape, " is not compatible with current shape: ", original_shape]
-        )
+    skip_shape_comparison = (
+        isinstance(original_param, (DefaultParameter, OptionalParameter)) and original_param.shape is None
+    )
+    if not skip_shape_comparison:
+        # We need to evaluate here anyway, so we map the entire shape to numbers upfront to save us from recomputing
+        # them again later.
+        original_shape = tuple(map(int, original_param.shape))
+        new_shape = tuple(map(int, new_param.shape))
+        if original_shape != new_shape:
+            is_compatible = utils.result.Result.err(
+                ["New parameter shape: ", new_shape, " is not compatible with current shape: ", original_shape]
+            )
 
-    # Once we know the concrete shape of the parameter, we can update the trace tensor accordingly.
-    # This not only makes the trace more informative, but is actually required for some APIs, like
-    # addConvolutionND (to set kernelDims correctly).
-    new_param.trace_tensor.shape = new_shape
+        # Once we know the concrete shape of the parameter, we can update the trace tensor accordingly.
+        # This not only makes the trace more informative, but is actually required for some APIs, like
+        # addConvolutionND (to set kernelDims correctly).
+        new_param.trace_tensor.shape = new_shape
```

Comment:
This part should be done unconditionally (the shape of the incoming parameter is always known)